src/families/cosmos/bridge/js.ts

```diff
@@ -0,0 +1,738 @@
+    !errors.amount &&
+    (validatorAmount.lt(0) || totalSpent.gt(a.spendableBalance))
+  ) {
+    errors.amount = new NotEnoughBalance();
+    totalSpent = new BigNumber(0);
+  }
+
+  return Promise.resolve({
+    errors,
+    warnings,
+    estimatedFees,
+    amount: new BigNumber(0),
+    totalSpent,
+  });
+};
+
+const prepareTransaction = async (a, t) => t;
+
+const signOperation = ({
+  account,
+  deviceId,
+  transaction,
+}: {
+  account: Account;
+  deviceId: any;
+  transaction: Transaction;
+}): Observable<SignOperationEvent> =>
+  withDevice(deviceId)((transport) =>
+    Observable.create((o) => {
+      let cancelled;
+
+      async function main() {
+        const { fees } = transaction;
+        if (!fees) throw new FeeNotLoaded();
+
+        const { freshAddress } = account;
+
+        const ledgerSigner = new LedgerSigner(transport, {
+          // todo: handle path derivation
+          hdPaths: [makeCosmoshubPath(0)],
+        });
+
+        o.next({ type: "device-signature-requested" });
+
+        let msgs;
+        switch (transaction.mode) {
+          case "send":
+            msgs = [
+              {
+                type: "cosmos-sdk/MsgSend",
+                value: {
+                  from_address: freshAddress,
+                  to_address: transaction.recipient,
+                  amount: [
+                    {
+                      denom: "atom",
+                      amount: `${transaction.amount.div(1000000)}`,
+                    },
+                  ],
+                },
+              },
+            ];
+
+            break;
+
+          case "delegate":
+            msgs = [
+              {
+                type: "cosmos-sdk/MsgDelegate",
+                value: {
+                  amount: [
+                    {
+                      amount: `${transaction.amount.div(1000000)}`,
+                      denom: "atom",
+                    },
+                  ],
+                  delegator_address: freshAddress,
+                  // todo:
+                  // validator_address: transaction.validator_address,
+                  validator_address: "",
+                },
+              },
+            ];
+
+            break;
+
+          case "undelegate":
+            // todo:
+            msgs = [{}];
+
+            break;
+
+          default:
+            throw "not supported";
+        }
+
+        const fee = {
+          amount: [
+            {
+              amount: `${transaction.fees?.div(1000000)}`,
+              denom: "atom",
+            },
+          ],
+          gas: `${transaction.gas}`,
+        };
+
+        const { accountNumber, sequence } = await getAccount(freshAddress);
+        const chainId = await getChainId();
+
+        const unsigned = {
+          chain_id: chainId,
+          account_number: accountNumber,
+          sequence: sequence,
+          fee: fee,
+          msgs: msgs,
+          memo: transaction.memo || "",
+        };
```

Comment:
You can extract the build of the unsigned payload in a dedicated `buildTransaction` function, that may be also used to compute fees.

Reply:
fix 94d5919